cluster-logging: Fine grained logs access (#1419)
openshift/enhancements: enhancements/cluster-logging/fine-grained-per-namespace-logs-access.md

The proposal provides an overview of the changes required to achieve this goal as well as some implementation details.

## Motivation

In enterprise environments, where OpenShift is used across different legal entities, it's common to have central teams that support the application teams in the respective entities. But given that some application may log sensitive data, those centralized support teams are not granted access to logs but they can only view specific objects, such as pods in the namespace.

Currently, access to logs in LokiStack is granted when a user has access to the given namespace or when the user is part of a specific cluster-admin Group. And even though OpenShift Container Platform 4 does allow to configure RBAC to address this issue, LokiStack does not, and therefore grants access to logs to people that should not see them.

This enhancement proposal presents a solution that enables cluster admins to have a more fine grained control over who accesses what logs.

### User Stories

> * Use Case number one, is where a user has access to a specific namespace to see the objects included but is denied to view logs from pods (for legal reason mostly).

Reviewer: Consider writing this like:

As an administrator,
I want to restrict visibility of certain logs for users with limited access
because there are legal reasons to do so.

... or something like that:

As what role
What we are doing
and why

* Use Case number two, is where a number of people have elevated permissions and therefore are able to access pretty much all namespaces on the OpenShift cluster. Usually those people are responsible for enabling and supporting applications when onboarding to OpenShift and therefore have permissions to see most of the objects in the given namespaces. However, due to legal and data protection constrains, those users can not have access to logs and therefore LokiStack should prevent them from seeing application specific logs.

### Goals

> * Cluster admins can grant and revoke access of the workload logs of specific namespaces to users.

Reviewer: Not sure if its additional goal or rewording but fundamentally we probably desire to make use of OpenShift RBAC in order that we are consistent with how permissions are managed elsewhere in lieu of  the contrived way we utilize now (e.g. Elasticsearch perms)

* Cluster admins can grant and revoke access of the workload logs of specific namespaces to users with elevated privileges.

### Non-Goals

* Prevent users with elevated privileges from escalating their privileges to access application logs that they are not supposed to have access to. 

> ## Proposal

Reviewer: I question the need to modify the CL api to support this features.  Do we need to explicitly enable fine grained access control?  I would propose by moving towards what I have [proposed](https://github.com/openshift/enhancements/pull/1410/files#diff-9d07913b066ab6b25695e5fcd8e023e05ef0761832052d53cc76cf826f699448R81-R91) for multi clusterlogforwarding you may be able to "enable" the feature by creating the right permissions set.

How is it loki determines access?  Will it be by executing a SAR? If so, then I propose we can be more clever how we define roles by utilizing virtual resources.

Reviewer 2: LokiStack is utilizing SARs against virtual resources already. Adding the switch is to keep backwards compatibility with the present implementation in CLO that gives all `system:authenticated` access to the logs by default. Of course it can be resolved without the switch proposed here by adding some upgrade code in CLO, but do users want to have full control of this update is not known upfront.

Reviewer: > Of course it can be resolved without the switch proposed here by adding some upgrade code in CLO, but do users want to have full control of this update is not known upfront

This is what they are used to now and would be OOTB going forward.  Either way we are giving a knob to an admin which is either our spec or a set of RBAC commands.  I would contend it is better to give them instructions on "opting out" by using RBAC in lieu of adding a knob which changes the behavior of the operator by adding some binary decision

```yaml
apiVersion: logging.openshift.io/v1
kind: ClusterLogging
metadata:
  name: instance
  namespace: openshift-logging
spec:
  logStore:
    type: lokistack
    lokistack:
      name: lokistack-dev
      advancedLogsAccess: true
  ...
```

2. The Cluster Logging Operator (CLO) deletes the `logging-all-authenticated-application-logs-reader` ClusterRoleBinding. This binding allows all authenticated users to see application logs in namespace where they have access.

3. The CLO then creates a new empty group `log-reader-group` and binds it to `logging-application-logs-reader`:

```yaml
apiVersion: user.openshift.io/v1
kind: Group
metadata:
  name: log-reader-group
users:
```
The group contains no users. The cluster admin can then add users to the group by editing it manually.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: logging-application-logs-reader-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: logging-application-logs-reader
subjects:
- kind: Group
  name: log-reader-group
  apiGroup: rbac.authorization.k8s.io
```

> 4. The operator also creates a new ClusterRole called `restricted-cluster-admin` and binds it to a new group `restricted-cluster-admin-group`. `restricted-cluster-admin` grants access to all api objects except for logs:

Reviewer 3: This step to me feels weird... it looks like we are encouraging people to use `restricted-cluster-admin` whilst in reality, IIUC we want to give them the tools to be able to restrict log access to pre-defined (by them) ClusterRoles. This to me would be done through good documentation on how to configure a privileged ClusterRole and how to grant it log access (by creating the RoleBinding `logging-application-logs-reader`). Please correct me if I'm miss understanding something

Author: The ClusterRole created here and its binding are more for convenience. We can go without them, but I think for the case where there are cluster admins with restricted access to logs, creating this would make it easier to configure out of the box.

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: restricted-cluster-admin
rules:
- apiGroups:
  - ''

>   - 'admissionregistration.k8s.io/v1'

Reviewer: What is the significance of these groups and how do they relate to viewing admin logs?  This looks like we have redefined an admin role that is a subset of something else.  Again, consider the use of virtual resources as I am [proposing](https://github.com/openshift/enhancements/pull/1410/files#diff-9d07913b066ab6b25695e5fcd8e023e05ef0761832052d53cc76cf826f699448R81-R91).  

* The operator can check if a user or group satisfies that role
* An administrator can create a role binding to any user or group
* Conceptionally we can apply the same logic to namespaced roles and bindings to restrict namespaces

  - '*'
  verbs:
  - '*'
```

```yaml
apiVersion: user.openshift.io/v1
kind: Group
metadata:
  name: restricted-cluster-admin-group
users:
```
The group contains no users. The cluster admin can then add users to the group by editing it manually.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: restricted-cluster-admin-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: restricted-cluster-admin
subjects:
- kind: Group
  name: restricted-cluster-admin-group
  apiGroup: rbac.authorization.k8s.io
```

Now the cluster admin can allow access to specific namespaces to the `restricted-cluster-admin-group` (or to a specific user in the group) by creating a RoleBinding with the `logging-application-logs-reader`, e.g.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: special-perm-restricted-cluster-admin-binding
  namespace: desired-namespace
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: logging-application-logs-reader
subjects:
- kind: Group # or User
  name: restricted-cluster-admin-group # or the User's name
  apiGroup: rbac.authorization.k8s.io
```

> All of the resources created will not be managed by the operator once they are created. This will allow the cluster admin to have control over how and when they are used.

Author: These resources are created for convenience purposes. They are not meant to be managed by the operator as this is just simple RBAC than can be handled by the admin.

Here's the modification for the `ClusterLogging` CRD:

```go
// LokiStackStoreSpec is used to set up cluster-logging to use a LokiStack as logging storage.
// It points to an existing LokiStack in the same namespace.
type LokiStackStoreSpec struct {
	// Name of the LokiStack resource.
	//
	// +required
	Name string `json:"name"`

	// Enable fine grained control over access to logs.
	//
	// +nullable
	// +optional
  AdvancedLogsAccess bool `json:"advancedLogsAccess"`
}
```

Here's an example of a `ClusterLogging` definition with advanced logs access enabled:

```yaml
apiVersion: logging.openshift.io/v1
kind: ClusterLogging
metadata:
  name: instance
  namespace: openshift-logging
spec:
  managementState: Managed
  logStore:
    type: lokistack
    lokistack:
      name: lokistack-dev
      advancedLogsAccess: true
  collection:
    type: vector
```

### Risks and Mitigations

- Privilege escalation, ie, a restricted cluster admin being able to grant themselves the ability to access logs they should not be able to access via an RBAC modification, is something cluster admins should be aware of and manage themselves.

Other risks TBD.

### Drawbacks

> This proposition assumes that once the advanced logs access is enabled and RBAC resources are created by the operator, the resources are no longer managed by the operator. This means that the remaining responsibilities, such as keeping the list of rules in `restricted-cluster-admin` up to date, will fall onto the cluster admins.

Reviewer: I believe by modifying our approach we defer these responsibilities entirely to who should control it, a cluster admin.  The operator may not need to modify any rbac

### Workflow Description

1. The Cluster Logging Operator (CLO) creates `cluster-logging-application-view`, `cluster-logging-infrastructure-view` and `cluster-logging-audit-view` ClusterRoles:

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: cluster-logging-application-view
rules:
- apiGroups:
  - loki.grafana.com
  resources:
  - application
  resourceNames:
  - logs
  verbs:
  - get
```

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: cluster-logging-infrastructure-view
rules:
- apiGroups:
  - loki.grafana.com
  resources:
  - infrastructure
  resourceNames:
  - logs
  verbs:
  - get
```

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: cluster-logging-audit-view
rules:
- apiGroups:
  - loki.grafana.com
  resources:
  - audit
  resourceNames:
  - logs
  verbs:
  - get
```

2. The cluster admin deletes the `logging-all-authenticated-application-logs-reader` ClusterBinding.

> 3. The cluster admin create the necessary binding to grant access to users, e.g.:

Author: `ProjectTemplate` can be used, but ultimately they create RBAC, so I'm keeping the proposal simple in that it only mentions RBAC and any automation is left to the admin to handle on their terms.

subjects:
- kind: User
  name: simple-user
  apiGroup: rbac.authorization.k8s.io
```

Granting `admin-user` access to `infrastructure` logs.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: admin-user-infrastructure-logs
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-logging-infrastructure-view
subjects:
- kind: User
  name: admin-user
  apiGroup: rbac.authorization.k8s.io
```

### API Extensions
N/A

### Implementation Details/Notes/Constraints

In the CLO, we need to disable the reconciliation of `logging-all-authenticated-application-logs-reader` ClusterRoleBinding. This reconciliation happens here:

`internal/logstore/lokistack/logstore_lokistack.go:64`
```go
func ReconcileLokiStackLogStore(k8sClient client.Client, deletionTimestamp *v1.Time, appendFinalizer func(identifier string) error) error {
  
  ...
  if err := reconcile.ClusterRoleBinding(k8sClient, lokiStackAppReaderClusterRoleBindingName, newLokiStackAppReaderClusterRoleBinding); err != nil {
    return kverrors.Wrap(err, "Failed to create or update ClusterRoleBinding for reading application logs.")
  }

  return nil
}
```

> ### Risks and Mitigations

Author: `This proposition assumes and is dependent on the fact that the SAR is implemented and works correctly.`

* Privilege escalation, ie, a restricted cluster admin being able to grant themselves the ability to access logs they should not be able to access via an RBAC modification, is something cluster admins should be aware of and manage themselves.

* This proposition assumes and is dependent on the fact that the SAR is implemented and works correctly.

> Other risks TBD.

Reviewer 3: nit
```suggestion

```

### Drawbacks

> * We don't automate much as this is an RBAC configuration focused proposal, so in order for cluster admins to properly configure access to logs, a comprehensive documentation is required.
> 
> Other drawbacks TBD.

Reviewer 3: nit
```suggestion
* RBAC configuration will not be managed by the operator, so in order for cluster admins to properly configure access to logs, comprehensive documentation is required on the subject.
```
